Updating mounting logic

arthurdm opened the pull request:
Updating the mounting logic of the spec.  In the last interlock we decided to mount all of the binding data, to allow for bindings projections into the container without a central coordinator.  

It may be easier to visualize the new text directly from the branch: https://github.com/application-stacks/service-binding-specification/tree/arthurdm-patch-1#mounting-and-injecting-binding-information

yharish991 commented on README.md:
````diff
@@ -177,44 +177,39 @@ Example of a partial CR:
+```
+
+Where:
+* `<mountPathPrefix>` defaults to `/platform/bindings` if not specified in the `ServiceBinding` CR via the `mountPathPrefix` element.  The environment variable `SERVICE_BINDINGS` **MUST** be set to the value of `<mountPathPrefix>`, so that applications can always find this directory.
````
would it nice if we rename the `SERVICE_BINDINGS` to `SERVICE_BINDINGS_MOUNT_PATH` or something similar which better implies the intent and meaning?

arthurdm replied:
good point @yharish991 - how about `SERVICE_BINDINGS_ROOT`?

navidsh commented on README.md:
````diff
@@ -177,44 +177,39 @@ Example of a partial CR:
+```
+
+Where:
+* `<mountPathPrefix>` defaults to `/platform/bindings` if not specified in the `ServiceBinding` CR via the `mountPathPrefix` element.  The environment variable `SERVICE_BINDINGS_ROOT` **MUST** be set to the value of `<mountPathPrefix>`, so that applications can always find this directory.
````
- What are your thoughts on making `SERVICE_BINDINGS_ROOT` available in `<mountPathPrefix>/<service-id>/metadata/envVars` in addition to env var?
- Should `SERVICE_BINDINGS_ROOT` be available as a key-value pair in the binding secret?
- Do you think the spec should mention anything about when there are multiple `BindingRequest` CRs with conflicting `mountPathPrefix`?

arthurdm replied:
I wonder if that would be redundant, since to reach `<mountPathPrefix>/<service-id>/metadata/envVars` the app would already know the value of `SERVICE_BINDINGS_ROOT`?   I think the pattern is:  the app code fetches `SERVICE_BINDINGS_ROOT` and then crawls through the bindings there.

arthurdm replied:
no harm in including it though

navidsh replied:
> ...since to reach `<mountPathPrefix>/<service-id>/metadata/envVars` the app would already know the value of `SERVICE_BINDINGS_ROOT`...

That's an important point that I missed lol... I agree that it's redundant and no point in adding

navidsh commented on README.md:
````diff
@@ -177,44 +177,39 @@ Example of a partial CR:
+```
+
+Where:
+* `<mountPathPrefix>` defaults to `/platform/bindings` if not specified in the `ServiceBinding` CR via the `mountPathPrefix` element.  The environment variable `SERVICE_BINDINGS_ROOT` **MUST** be set to the value of `<mountPathPrefix>`, so that applications can always find this directory.
````
Should `SERVICE_BINDINGS_ROOT` be available as a key-value pair in the binding secret?

arthurdm replied:
that's a good point. I think generally there's an implementation issue to be solved here:  if one entity (such as the Service Binding Operator) is processing the `ServiceBinding` CR and creating the target Secret, how does another entity (such as the Runtime Component Operator) know things such as `SERVICE_BINDING_ROOT` as well as each `<service-id>` to use.  

I think that means this second entity also needs to look into the `ServiceBinding` CR.  But another general issue is:  what about `dataMappings`, which `<service-id>` do we associate it with? CC  @sbose78

navidsh commented on README.md:
````diff
@@ -177,44 +177,39 @@ Example of a partial CR:
+```
+
+Where:
+* `<mountPathPrefix>` defaults to `/platform/bindings` if not specified in the `ServiceBinding` CR via the `mountPathPrefix` element.  The environment variable `SERVICE_BINDINGS_ROOT` **MUST** be set to the value of `<mountPathPrefix>`, so that applications can always find this directory.
````
Do you think the spec should mention anything about when there are multiple `BindingRequest` CRs with conflicting `mountPathPrefix`?

arthurdm replied:
yes, good point!

arthurdm replied:
I added a new commit to clarify that the value of `SERVICE_BINDINGS_ROOT` is immutable.

arthurdm commented:
@navidsh / @sbose78 - I added to the PR a new section [here](https://github.com/application-stacks/service-binding-specification/blob/arthurdm-patch-1/README.md#cross-services-data-mappings) to deal with `dataMappings` as a special cross-service id, and another section [here](https://github.com/application-stacks/service-binding-specification/blob/arthurdm-patch-1/README.md#guidelines-for-intermediate-binding-representation) about an intermediate representation of the Secret which SBO could produce and other Operators (such as Runtime Component Operator, or others) could consume.

navidsh commented on README.md:
```diff
@@ -177,44 +177,65 @@ Example of a partial CR:
+
+#### Cross services data mappings
+
+The `ServiceBinding` CR allows for the declaration of mappings that are potentially composed of different service bindings (ids).  For this case, the special id of `dataMappings` is reserved and **MUST NOT** be used by any service bindings.  
```
`dataMappings` vs `data-mappings`?
I couldn't find an official Linux naming convention guideline. But overall it seems like using lower-case/dashe is more common. e.g. https://developers.google.com/style/filenames

navidsh replied:
Same goes for `envVars`

arthurdm replied:
that's a good question.  I guess for filenames that's more common, but should we view these as traditional files or more like k8s bindings?  My view is that yes they are files, but probably not in the sense of regular r/w OS files. 

In the binding schema we have `basePath` and `bindingRole`, which would become "files" in the persisted model.

yharish991 replied:
this is a good conversation, it looks like its more common in Linux file systems to have files with `-` to separate logical words. 

Cloud native buildpack specification also uses `-` as the separator here in the example https://github.com/buildpacks/spec/blob/master/extensions/bindings.md#example-directory-structure-1 and here it talks about the naming convention https://github.com/buildpacks/spec/blob/master/extensions/bindings.md#launch-phase

I think we should probably have a section that talks about naming convention, else we might have files that are in different naming formats. The spec also allows extra binding properties, if either the backing operator service or the custom name in the dataMappings portion chooses to use `-` in the binding name, then we might have few files which are using camel case and few with `-`. I think we should have consistent naming convention.

navidsh replied:
I personally prefer `dash-delimited` over `camelCase`. It looks more familiar in Kubernetes/container.
e.g. [K8s resource naming convention](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names):
```
a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')" for field "metadata.name".
```

And the secret bindings generated by SBO will be following the k8s naming rules.
Having said that I'm good either way 🙂 

+1 for a naming convention (eventually).

arthurdm replied:
good point guys.  I will update the PR, and will open a separate issue to go through the whole spec and apply the name convention change (plus a notice about it).

yharish991 commented on README.md:
```diff
@@ -177,44 +177,65 @@ Example of a partial CR:
+
+#### Cross services data mappings
+
+The `ServiceBinding` CR allows for the declaration of mappings that are potentially composed of different service bindings (ids).  For this case, the special id of `dataMappings` is reserved and **MUST NOT** be used by any service bindings.  
```
this is a comment from implementation perspective, right now in the `ServiceBinding` CR,  `customEnvVar` is global for all backing services, what if we have `customEnvVar` per backing service ? that way we can completely avoid the scenario of cross services data mappings and unnecessary complexity for the application to go to multiple paths to get the binding information of one single service.

From this
```
backingServiceSelectors:
    - group: postgres.dev
      kind: Service
      resourceRef: user-db
      version: v1beta1
      id: X
    - group: postgres.dev
      kind: Credentials
      resourceRef: admin-creds
      version: v1beta1
      id: Y
customEnvVar:
    - name: CONNECTION_URL
       value: {{  X.status.username }} / {{ Y.status.url }}
```

to this 

```
backingServiceSelectors:
    - group: postgres.dev
      kind: Service
      resourceRef: user-db
      version: v1beta1
      id: X
    - group: postgres.dev
      kind: Credentials
      resourceRef: admin-creds
      version: v1beta1
      id: Y
      customEnvVar:
         - name: CONNECTION_URL
         value: {{  X.status.username }} / {{ Y.status.url }}
```
Service binding CR yaml snippet extracted from here https://github.com/redhat-developer/service-binding-operator/issues/396

Thoughts @arthurdm @navidsh @sbose78 ?

navidsh replied:
@yharish991 Yea, I think that would simplify things a lot but then there is this concern:
https://github.com/redhat-developer/service-binding-operator/issues/356#issuecomment-596049739

`CONNECTION_URL` is technically relying on info from service (`id: X`) but is defined under another service. But in reality, it relies on both services to be defined in the `ServiceBinding` CR.

arthurdm replied:
ya.  I think it's a good thought @yharish991 , but as @navidsh pointed out it is counter intuitive to have something under one service relying on another.  I modified the PR to call the special id `custom-mappings`, so it's very clear this is something custom / composed by the requester.

yharish991 replied:
In SQl, its quite common to refer to multiple tables all over the query, so was thinking why not here, the developers are already used to this kind of thinking, so i think it would not be a problem to them, also it solves the problem of identifying the binding info if we use `custom-mappings` as the special id. For example if the directory `<mountPathPrefix>/custom-bindings/secret/` has file `CONNECTION_URL `, how do we identify which service it belongs to?

arthurdm replied:
I think that when a custom mapping is used, if it refers to a particular service it should use a reasonable prefix?

navidsh commented on README.md:
````diff
@@ -177,44 +177,65 @@ Example of a partial CR:
+    ```
+  * `<service-id>.metadata.<metadata-name>: <metadata-value>`, representing a single metadata item.  Example:  
+    ```
+    mongo-db/metadata/envVars: |-
````
```suggestion
    mongo-db.metadata.envVars: |-
```

navidsh commented on README.md:
````diff
@@ -177,44 +177,65 @@ Example of a partial CR:
+* a single Secret, whose name matches the corresponding `ServiceBinding` CR's `metadata.name` and resides in the same namespace.  Each item inside this Secret is in the form of either:
+  * `<service-id>.secret.<binding-name>: <binding-value>`, representing a single binding item. Example: 
+    ```
+    mongo-db/secret/MONGODB_HOST: myhost.com
````
```suggestion
    mongo-db.secret.MONGODB_HOST: myhost.com
```

arthurdm commented:
Updated the PR to use dash separated name and created https://github.com/application-stacks/service-binding-specification/issues/42 to track future consistency in this space.

navidsh commented on README.md:
```diff
@@ -177,44 +177,64 @@ Example of a partial CR:
+However, it is **RECOMMENDED** that implementations expose their intermediate binding representation (i.e. the model containing the binding data prior to mount) in the following way:
+
+* a single Secret, whose name matches the corresponding `ServiceBinding` CR's `metadata.name` and resides in the same namespace.  Each item inside this Secret is in the form of either:
+  * `<service-id>.secret.<binding-name>: <binding-value>`, representing a single binding item. Example: 
```
@arthurdm what are your thoughts on including `type` info from [Service Binding Schema](https://github.com/application-stacks/service-binding-specification#service-binding-schema)?
e.g. `<service-id>.<type>.secret.<binding-name>: <binding-value>`

So app developers can easily query what data is available for particular type e.g. `db`, `kafka`

Alternatively, it is possible to just rely on `<service-id>.metadata.<metadata-name>: <metadata-value>` for convey that information e.g. `<service-id>.metadata.type: kafka` but it's just more work for a dev to find what `service-id` is for a particular `type`

yharish991 replied:
this definitely solves this [problem](https://github.com/application-stacks/service-binding-specification/pull/41#discussion_r430713400) , but add complexity for the application developer to find what `service-id` is for a particular type.

arthurdm replied:
that's an interesting question.  At the moment `type` is part of the binding schema, so we're advocating for it to be inside the `secret` path.  But it does feel more natural to be in `metadata`.  Something we'll need to solve.  I would vote for keeping the type out of the path, because I am not sure there's a reasonable default.

scothis replied:
The CNB Bindings spec defines two required bits of metadata `kind` and `provider`. The kind is particularly important at runtime so an application can consume a service by kind instead of by name.

arthurdm replied:
yeah - we'll probably need another future PR for the recommended set of metadata.

arthurdm commented on README.md:
````diff
@@ -177,44 +177,65 @@ Example of a partial CR:
-  ]
-}
-```
+If a particular binding key is made available as an environment variable an entry **MUST** be made at the following file:
````
The intent is to enable both scenarios:
(1) when the entity that is projecting the bindings is also the one that is making the env var available
(2) when some runtime library is walking over the mounted bindings and creating env vars for some of the bindings (presumably during app startup)

In both cases the env vars are available for the application to consume, and there's a consistent way that describes which env vars are available (agnostic to how things were projected or processed).

nebhale replied:
I'm very wary of standardizing how the canonical representation is transformed into environment variables or even whether it is.  At the very least, I don't think this should be a part of the core spec, but rather an extension specification defined as "this is optional functionality that not all platforms need to support, but if they do this is how it should be done".

nebhale replied:
In general, I consider this functionality a convenience and conventions should emerge over time as the community decides how they like describing this functionality rather than prescribing it up front.  For example, in many of the Cloud Native Buildpacks that currently exist, we wouldn't use this file as [we describe a convention](https://github.com/paketo-buildpacks/azure-application-insights/blob/master/properties/properties.go#L44-L48) like the following:

```
for k, v := range b.Secret {
  s := strings.ToUpper(k)
  s = strings.ReplaceAll(s, "-", "_")
  s = strings.ReplaceAll(s, ".", "_")

  vars = append(vars, fmt.Sprintf(`export APPINSIGHTS_%s="%s"`, s, v))
}
```

This convention-based approach means that we can map *every* secret, regardless of whether it's been defined in a JSON-structure to an environment variable giving flexibility to the provider (especially in the case where that provider is a user manually generating the binding)

arthurdm commented on README.md:
```diff
@@ -177,44 +177,65 @@ Example of a partial CR:
+
+#### Cross services data mappings
+
+The `ServiceBinding` CR allows for the declaration of mappings that are potentially composed of different service bindings (ids).  For this case, the special id of `dataMappings` is reserved and **MUST NOT** be used by any service bindings.  
```
That's right.  I added a new commit to use `<service-id>` for clarity.   Good point @scothis

arthurdm commented on README.md:
```diff
@@ -177,44 +177,65 @@ Example of a partial CR:
+
+This specification does not mandate a particular methodology for implementations to process `ServiceBinding` CRs and mount the resulting data as per the requirements set in the previous sections, which allows for different frameworks to have flexibility in exactly how they fulfill the binding request.
+
+However, it is **RECOMMENDED** that implementations expose their intermediate binding representation (i.e. the model containing the binding data prior to mount) in the following way:
```
It was in my head, but wasn't written down 😄 

I added a new commit which hopefully clarifies:
```
However, it is RECOMMENDED that implementations expose their intermediate binding representation (i.e. the model containing the binding data prior to mount) in the following way, so that Operator-based deployments or any other processing unit that owns the deployment can take over the responsibility for mounting the data into the container:
```

arthurdm replied:
Basically, there are scenarios where an Operator created the deployment, so if another Operator does the Service Binding processing we want a way to just fetch the resulting / processed Secret, and do the mounting via the first Operator., to avoid a reconcile call that will remove the mount.

nebhale commented on README.md:
```diff
@@ -177,44 +177,64 @@ Example of a partial CR:
 ### Mounting and injecting binding information
 
-This specification allows for data to be mounted using volumes or injected using environment variables.  The best practice is to mount any sensitive information, such as passwords, since that will avoid accidentally exposure via environment dumps and subprocesses.  Also, binding binary data (e.g. .p12 certificate for Kafka) as an environment variable might cause a pod to fail to start (stuck on ` ...
+#### Mounting data
```
When describing mounting of bindings, I think it's clearer to think about and describe a binding has having a name and two parts, the metadata and secret.  If you think about it this way, the flow of this section of the document talks first about `<mountPathPrefix>/<service-id>` and the two file-per-entry maps underneath it.

arthurdm replied:
that's a good approach @nebhale.  One thing I am wondering though, is if `metadata` is a required part, or a (strongly) recommended part.  

A consumer can certainly "get by" with just the `secret`, but with limitations that `metadata` could help unlock (like switching to another compatible binding of the same kind).   

In terms of the spec I was thinking `secret` is a MUST and `metadata` is RECOMMENDED.  What do you think?

nebhale commented on README.md:
````diff
@@ -177,44 +177,64 @@ Example of a partial CR:
+```
+
+Where:
+* `<mountPathPrefix>` defaults to `/platform/bindings` if not specified in the `ServiceBinding` CR via the `mountPathPrefix` element.  The environment variable `SERVICE_BINDINGS_ROOT` **MUST** be set to the value of `<mountPathPrefix>`, so that applications can always find this directory, and it is immutable.  This means that if another `ServiceBinding` CR wants to project itself into the same co ...
````
I'd be very careful defining a default value for this (we got some pushback on it), but if you do I'd also choose something shallower.  The CNB project chose `/platform/bindings` (and only at build time, not run time) because there are so many moving parts in the spec, that it made sense to have an umbrella directory where `bindings` is a peer to other kinds of data.  Because this spec is concerned with a single data type, bindings, the default (if it exists) should probably be a top-level directory, `/bindings`.

nebhale replied:
Also, I believe that both having a default and requiring `$SERVICE_BINDINGS_ROOT` are in opposition to one another.  If the environment variable is always required, then no default should be necessary and vice-versa.

arthurdm replied:
That's a good point about `/bindings`.  I must say that `/platform/bindings` was heavily influenced by the desire to stay consistent with what CNB had done.  

The `$SERVICE_BINDINGS_ROOT` is an interesting one:   from the app perspective, it can rely that it will always be set.   So the platform must always set it, even when the author of the binding did not specify a `mountPathPrefix`, thus the platform would need a default to use in such cases.

scothis replied:
The default should only exist within the reconciler attaching the binding to the workload. The env var should always be defined for a container that has any service bindings. In short, a workload can define the var to instruct the bindings to be mounted in a specific location, otherwise the first reconciler to bind into the workload gets to define the path (to the default value).

nebhale commented on README.md:
```diff
@@ -177,44 +177,64 @@ Example of a partial CR:
+* `<service-id>` is the `id` field of the corresponding `service` entry in the `ServiceBinding` CR.  If the `id` field is not present, the `name` field is used instead.  The `<service-id>` path **MUST** be unique between the services bound to a particular application.
+* `<persisted_secret>` represents a set of files where the filename is a Secret key and the file contents is the corresponding value of that key.
+
+Example:  `/platform/bindings/mongo-db/secret/MONGODB_HOST`
```
A better example here wouldn't include the "kind" in the service id even if this is how it would be practically used.  You should also be thinking up front about the possibility that bindings aren't obviously environment variables.  How about `/platform/bindings/my-nosql-db/secret/grid-fs-database`?  This is actually a real-life binding for Mongo, but illustrates how the names of files on the filesystem don't immediately identify things.

nebhale replied:
Note, this suggestion to remove the Mongo-obvious names, while retaining Mono-semantic names applies throughout the document.

arthurdm replied:
good point about `mongo-db`.  I will switch to something that doesn't suggest this field is the `kind`.  

Yes, definitely we will have very `ugly` binding names.   This is where the `dataMappings` / `custom mapping` conversation comes in.  So we will have the ugly names, as well as the "environment var friendly" name.  😄

nebhale commented on README.md:
```diff
@@ -177,44 +177,64 @@ Example of a partial CR:
+
+Example:  `/platform/bindings/mongo-db/secret/MONGODB_HOST`
+
+#### Custom bindings
```
I don't believe that this needs to be reserved for custom usage.  If you think about the projection of a binding into the container as being loosely coupled to what has provided the metadata/secret for it, then all of the binding ids are equivalent and any of them could be "custom"; the application in the container would never need to know.  A binding projection could be:

* a perfect copy of a binding outside the container
* a copy of a binding outside the container with the name changed
* a completely custom configuration with a `ConfigMap` and `Secret` defined on a `Deployment` (just handwritten into a Spec)
* a custom configuration that makes reference to other `ConfigMap`s and `Secret`s some hand written, some from bindings outside the container.

I think this section should be removed as it's a special case of something that doesn't need to be special-cased.

arthurdm replied:
yeah, this is definitely an area where we can give more thought.    

Historically, our flow was that all of the service bindings that the app "sees" are:
(1) the bindings we fetched from the services defined in the `ServiceBindingRequest` CR 
(2)the customized `dataMappings`, which included renamed items as well as cross-service composed bindings.

They were always persisted as a single list of bindings.  

Now that we have switched to a model where the bindings are persisted in a path that includes the `service-id`, we have this issue to solve where the `dataMappings` elements from our `ServiceBindingRequest` CR can be renames of multiple services.  

One suggestion that @yharish991 had is to move the `dataMappings` underneath each of the `services` entry - which would solve the problem on "where" we persist that (i.e. under the corresponding `service-id`.   But we still have issues with composed bindings (aggregating from multiple services).  Where would that be persisted?

nebhale replied:
I think there's a conflation of the actors involved here.  From the view of the projection as represented on the PodSpec, there are just "some volumes":

```yaml
spec:
  containers:
  - name: app
    image: nginx
    env:
    - name: SERVICE_BINDING_ROOT
      value: /platform/bindings
    volumeMounts:
    - name: my-nosql-db-metadata
       mountPath:  /platform/bindings/my-nosql-db/metadata
       readOnly: true
    - name: my-nosql-db-secret
       mountPath:  /platform/bindings/my-nosql-db/secret
       readOnly: true
  volumes:
  - name: my-nosql-db-metadata
     configMap:
       name: my-mongodb-metadata
  - name: my-nosql-db-secret
     secret:
       secretRef: my-mongodb-secret
```

How those volumes got reconciled to there, what their names are and how those were calculated, don't really matter in the context of this part of the spec.  What makes this so nice is that *another* reconciler can come along and do something fun like adding a binding that is synthesized from other bindings.

```yaml
spec:
  containers:
  - name: app
    image: nginx
    env:
    - name: SERVICE_BINDING_ROOT
      value: /platform/bindings
    volumeMounts:
    - name: my-nosql-db-metadata
       mountPath:  /platform/bindings/my-nosql-db/metadata
       readOnly: true
    - name: my-nosql-db-secret
       mountPath:  /platform/bindings/my-nosql-db/secret
       readOnly: true

    - name: my-rdbms-metadata
       mountPath:  /platform/bindings/my-rdbms/metadata
       readOnly: true
    - name: my-rdbms-secret
       mountPath:  /platform/bindings/my-rdbms/secret
       readOnly: true
  volumes:
  - name: my-nosql-db-metadata
     configMap:
       name: my-mongodb-metadata
  - name: my-nosql-db-secret
     secret:
       secretRef: my-mongodb-secret

  - name: my-rdbms-metadata
     configMap:
       name: my-custom-metadata
  - name: my-rdbms-secret
     secret:
       secretRef: my-custom-secret

---
apiVersion: v1
kind: ConfigMap
metadata:
  name: my-custom-metadata
data:
  kind: mysql
---
apiVersion: v1
kind: Secret
metadata:
  name: my-custom-secret
type: Opaque
stringData:  # This data is synthesized by some other reconciler
  jdbcUrl: jdbc:mysql://${mysql.secret.host}:${mysql.secret.port}/${mysql.secret.database}?reconnect=true
  password: ${mysql.secret.password}
  username: ${mysql.secret.username}
```

Key here though, is that this isn't a spec at all.  It's just another pair of volumes, mounted to a predictable location, by some other reconciler (essentially a distributed model of binding generation).

Finally, there's nothing to stop a user from *just creating a hand-rolled-binding*:

```yaml
spec:
  containers:
  - name: app
    image: nginx
    env:
    - name: SERVICE_BINDING_ROOT
      value: /platform/bindings
    volumeMounts:
    - name: my-nosql-db-metadata
       mountPath:  /platform/bindings/my-nosql-db/metadata
       readOnly: true
    - name: my-nosql-db-secret
       mountPath:  /platform/bindings/my-nosql-db/secret
       readOnly: true

    - name: my-rdbms-metadata
       mountPath:  /platform/bindings/my-rdbms/metadata
       readOnly: true
    - name: my-rdbms-secret
       mountPath:  /platform/bindings/my-rdbms/secret
       readOnly: true

    - name: my-new-relic-metadata
       mountPath:  /platform/bindings/my-new-relic/metadata
       readOnly: true
    - name: my-new-relic-secret
       mountPath:  /platform/bindings/my-new-relic/secret
       readOnly: true
  volumes:
  - name: my-nosql-db-metadata
     configMap:
       name: my-mongodb-metadata
  - name: my-nosql-db-secret
     secret:
       secretRef: my-mongodb-secret

  - name: my-rdbms-metadata
     configMap:
       name: my-custom-metadata
  - name: my-rdbms-secret
     secret:
       secretRef: my-custom-secret

  - name: my-new-relic-metadata
     configMap:
       name: my-hand-rolled-metadata
  - name: my-new-relic-secret
     secret:
       secretRef: my-hand-rolled-secret
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: my-custom-metadata
data:
  kind: mysql
---
apiVersion: v1
kind: Secret
metadata:
  name: my-custom-secret
type: Opaque
stringData:  # This data is synthesized by some other reconciler
  jdbcUrl: jdbc:mysql://${mysql.secret.host}:${mysql.secret.port}/${mysql.secret.database}?reconnect=true
  password: ${mysql.secret.password}
  username: ${mysql.secret.username}

---
apiVersion: v1
kind: ConfigMap
metadata:
  name: my-hand-rolled-metadata
data:
  kind: new-relic
---
apiVersion: v1
kind: Secret
metadata:
  name: my-hand-rolled-secret
type: Opaque
stringData:  # This data is hand entered by a user
  licenseKey: <NEW-RELIC-LICENSE-KEY>
```

arthurdm replied:
awesome samples @nebhale!  thanks.  We should add some of these to the spec.  :)

arthurdm commented:
@nebhale - thanks so much for your comments!  They are really helpful and shed a different light on these concepts.

For some reason GitHub is not showing me a proper "reply" input for to the comments about the environment variable location, so adding them here.

I hear what you're saying in terms of different platforms having different conventions for creating environment variables, and I totally agree with it.   I think a reasonable goal though, is to try to bring consistency at least in terms of how an app finds out _which_ environment variables are available - in a platform-agnostic way.  So this is our attempt at something more robust than `VCAP_SERVICES`, but with similar benefits.

So using the example you provided, a framework could definitely define the set of `APPINSIGHTS_` env vars - all that the spec is mandating is that the names of these env vars are persisted into the corresponding `$SERVICE_BINDINGS_ROOT/<service-id>/metadata/envVars` file.        Do you think that is still too restrictive from a spec point of view?

nebhale commented:
I would think that's a bit too restrictive.  Ignoring this specification, how would an application know what environment variables are available to it?

1.  Pre-knowledge.  The New Relic agent knows to look for `$NEW_RELIC_LICENSE_KEY` and if it doesn't exist, well that's fine.
2.  Iteration.  Spring Boot lists all environment variables, turns them into properties and sets the values of those properties.

I don't think listing a subset of the environment variables available within a container (remember, users, OS, and images are setting a mountain of other env vars), in a file gains you much understandability about the environment you're running in.

arthurdm commented:
yeah that's true @nebhale - when I think about it, I can't find a good scenario where an app would rely on dynamically discovering env vars available to it from a persisted location, specially if it's just a subset of the vars.

arthurdm commented:
hey everyone - the last commits to the PR changed a few things:
- added a new sample CR to the `ServiceBinding` section, so that it is easy to illustrate the `dataMappings` per service
- added a new `EXPERIMENTAL` tagged section about synthetic bindings (essentially replacing the global `dataMappings` that we had)
- removed the optional intermediate Secret representation for now, as that was getting very messy in terms of propagating all the needed information to properly bind the secret and metadata per service.
